agent/api/container.go

```diff
@@ -246,6 +248,33 @@ func (c *Container) GetKnownExitCode() *int {
 	return c.knownExitCode
 }
 
+// SetRegistryAuthCredentials sets the credentials for pulling image from ECR
+func (c *Container) SetRegistryAuthCredentials(credential credentials.IAMRoleCredentials) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.RegistryAuthentication.ECRAuthData.SetPullCredentials(credential)
+}
+
+// ShouldPullWithExecutionRole returns whether this container has its own ECR credentials
+func (c *Container) ShouldPullWithExecutionRole() bool {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.RegistryAuthentication != nil &&
+		c.RegistryAuthentication.Type == "ecr" &&
+		c.RegistryAuthentication.ECRAuthData != nil &&
+		c.RegistryAuthentication.ECRAuthData.UseExecutionRole
+}
+
+// ShouldWaitForExecutionCredentials checks whether this container needs to wait for
+// credentials to pull form ECR
+func (c *Container) ShouldWaitForExecutionCredentials() bool {
```

Comment:
I feel like this function could just exist in the dependencygraph package, and then you don't need separate handling in engine for this specifically.  That gets this slightly out of being special-cased and handled like all the other dependencies.

Reply:
I thought about making this part in the dependencygraph package, but that will require the `credentials manager` passed into the `DependenciesAreResolved`, if you are ok with that, then I can make the change.

Reply:
Can't you just do it like this?

```go
// dependencygraph/graph.go

func credentialsAreResolved(target *api.Container) bool {
	if !target.ShouldPullWithExecutionRole() {
		return true
	}
	return target.GetKnownStatus() < ContainerPulled &&
		target.GetDesiredStatus() != ContainerStopped
}
```

Reply:
Ok, gotcha, so actually you are suggesting only move the container check into dependencygraph, while keeps the `isCredentialsResolved` in the engine.

Reply:
No, see my edit above.  I think you should move `isCredentialsResolved` into `dependencygraph`.